src/sources/pulsar.rs

```diff
@@ -0,0 +1,451 @@
+
+    /// Name of the dead topic where the failing messages will be sent.
+    pub dead_letter_topic: String,
+}
+
+struct FinalizerEntry {
+    consumer: Arc<Mutex<Consumer<String, TokioExecutor>>>,
+    message: pulsar::consumer::Message<std::string::String>,
+}
+
+impl std::fmt::Debug for FinalizerEntry {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("FinalizerEntry")
+            .field("message", &self.message.payload)
+            .finish()
+    }
+}
+
+impl GenerateConfig for PulsarSourceConfig {
+    fn generate_config() -> toml::Value {
+        toml::from_str(
+            r#"
+            topics = ["topic1", "topic2"]
+            endpoint = "pulsar://127.0.0.1:6650""#,
+        )
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+impl SourceConfig for PulsarSourceConfig {
+    async fn build(&self, cx: SourceContext) -> crate::Result<super::Source> {
+        let consumer = create_consumer(self).await?;
+        let decoder = DecodingConfig::new(
+            self.framing.clone(),
+            self.decoding.clone(),
+            LogNamespace::Legacy,
+        )
+        .build();
+        let acknowledgements = cx.do_acknowledgements(self.acknowledgements);
+
+        Ok(Box::pin(pulsar_source(
+            consumer,
+            decoder,
+            cx.shutdown,
+            cx.out,
+            acknowledgements,
+        )))
+    }
+
+    fn outputs(&self, _global_log_namespace: LogNamespace) -> Vec<Output> {
+        vec![Output::default(self.decoding.output_type())]
+    }
+
+    fn can_acknowledge(&self) -> bool {
+        true
+    }
+}
+
+async fn create_consumer(
+    config: &PulsarSourceConfig,
+) -> crate::Result<pulsar::consumer::Consumer<String, TokioExecutor>> {
+    let mut builder = Pulsar::builder(&config.endpoint, TokioExecutor);
+
+    if let Some(auth) = &config.auth {
+        builder = match (
+            auth.name.as_ref(),
+            auth.token.as_ref(),
+            auth.oauth2.as_ref(),
+        ) {
+            (Some(name), Some(token), None) => builder.with_auth(Authentication {
+                name: name.clone(),
+                data: token.inner().as_bytes().to_vec(),
+            }),
+            (None, None, Some(oauth2)) => {
+                builder.with_auth_provider(OAuth2Authentication::client_credentials(OAuth2Params {
+                    issuer_url: oauth2.issuer_url.clone(),
+                    credentials_url: oauth2.credentials_url.clone(),
+                    audience: oauth2.audience.clone(),
+                    scope: oauth2.scope.clone(),
+                }))
+            }
+            _ => return Err(Box::new(pulsar::error::Error::Authentication(
+                AuthenticationError::Custom(
+                    "Invalid auth config: can only specify name and token or oauth2 configuration"
+                        .to_string(),
+                ),
+            ))),
+        };
+    }
```

Comment:
I believe a bunch of this is shared with the sink as well. This particular bit could probably actually be a method on the auth struct.